### PR TITLE
Adding a function to wait for publisher confirms

### PR DESCRIPTION
**Pull request by manchicken**
- Also added an example program to demo it
- Also removed a deprecated uninit function call from the ssl example

### CODE REVIEW

**Comment by manchicken** [comment]
I'll sort out the example issue (llu vs Lu) later, but could you let me know if this approach works for y'all?

**Comment by manchicken** [comment]
I'm working on adding publisher confirms to the `Net::AMQP::RabbitMQ` Perl module, and this seemed like a bit of functionality I could add back upstream. Let me know if I missed anything, or if there's anything you'd like me to update.

The example isn't the most imaginative, I know, but it shows how to use the new function.

**Comment by manchicken** [comment]
@alanxz, it kinda looks like the windows builds are breaking on `master`? Do you have any advice on how to get that working? I'll fix the linting issue.

**Comment by manchicken** [comment]
Apologies for all of the confusion and CI runs, I don't have a Windows test environment. I think I got all CI passing now.

**Comment by alanxz** [comment]
Approach LGTM.

Will do a pass on the actual implementation shortly.

**Comment by manchicken** [comment]
Thanks! My C is a bit rusty, so feel free to point out anything I could improve there as well.

File: include/rabbitmq-c/amqp.h [hunk 1]
```diff
@@ -2433,6 +2433,26 @@ AMQP_EXPORT
...
+ *
+ * \param [in] state connection state
+ * \param [out] pointer to where the ACK details should go
+ * \param [in] timeout when waiting for the frame. Passing NULL will result in
+ * blocking behavior
+ * \returns amqp_rpc_reply_t *
+ */
+AMQP_EXPORT
+amqp_rpc_reply_t AMQP_CALL amqp_publisher_confirm_wait(
+    amqp_connection_state_t state, amqp_envelope_t *envelope,
```
Comment (alanxz) [inline include/rabbitmq-c/amqp.h:2453]: Could you describe the intended purpose of the envelope parameter?

Seems like this may not be required.

File: include/rabbitmq-c/amqp.h [hunk 2]
```diff
@@ -2433,6 +2433,26 @@ AMQP_EXPORT
...
+    amqp_basic_ack_t *ack, const struct timeval *timeout);
```
Comment (alanxz) [inline include/rabbitmq-c/amqp.h:2454]: "out" parameters should go after any "in" parameters (so this should go after timeout).

Alternatively consider making this function return `amqp_basic_ack_t*`, returning NULL on error, and making the caller consult amqp_get_rpc_reply(...) when NULL is returned. (I haven't thought through the ergonomics of an API like this, what you have may be preferable).

File: examples/amqp_ssl_connect.c [hunk 3]
```diff
@@ -95,7 +95,6 @@ int main(int argc, char const *const *argv) {
   die_on_amqp_error(amqp_connection_close(conn, AMQP_REPLY_SUCCESS),
                     "Closing connection");
   die_on_error(amqp_destroy_connection(conn), "Ending connection");
-  die_on_error(amqp_uninitialize_ssl_library(), "Uninitializing SSL library");
```
Comment (alanxz) [inline examples/amqp_ssl_connect.c:98]: Why is this change required?

**Review by alanxz — CHANGES_REQUESTED** [review]

File: include/rabbitmq-c/amqp.h (hunk 2 above)
Comment (alanxz) [reply include/rabbitmq-c/amqp.h:2454]: The other RPC-like methods use `amqp_simple_rpc_decoded`, which you can use for inspiration if this ends up being the preferred way of implementing this: https://github.com/alanxz/rabbitmq-c/blob/a2d0be96d6902fc909f2cc9a80fb385b4f9362c9/librabbitmq/amqp_socket.c#L1110 to retur

File: include/rabbitmq-c/amqp.h (hunk 1 above)
Comment (manchicken) [reply include/rabbitmq-c/amqp.h:2453]: Yeah, it's the only thing that'll tell you which channel the ACK came through. The code for the `amqp_basic_ack_t` is correct per spec, naturally, but it can be helpful to know which channel the ACK came on.

File: examples/amqp_ssl_connect.c (hunk 3 above)
Comment (manchicken) [reply examples/amqp_ssl_connect.c:98]: Deprecation notices complaining when building examples.

File: include/rabbitmq-c/amqp.h (hunk 2 above)
Comment (manchicken) [reply include/rabbitmq-c/amqp.h:2454]: I'll take a whack at this, thanks.

Comment (manchicken) [reply include/rabbitmq-c/amqp.h]: Yeah, I still kinda like my approach better for a few reasons:
- The ACK belongs to the library, so the application can't use the `amqp_basic_ack_t*` that is decoded. By having the application allocate their own and then put the data there lets the application own the memory there which seems like it'll make using this in threading more straight-forward.
- For some reason, I see this as more analogous to `amqp_consume_message()` than `amqp_simple_rpc_decoded()`. This could be some misunderstanding on my part, and I'd welcome a different perspective.

I've pushed changes for the parameter ordering piece, let me know if you have any additional tweaks you'd like to see.

**Comment by manchicken** [comment]
@alanxz Let me know if there's anything else you'd like to see changed here. I'd really like to get this into my Perl module, users have been asking for it for a long time.

File: include/rabbitmq-c/amqp.h (hunk 1 above)
Comment (manchicken) [reply include/rabbitmq-c/amqp.h]: Did you have any more questions or requests for change on this?

File: include/rabbitmq-c/amqp.h (hunk 2 above)
Comment (manchicken) [reply include/rabbitmq-c/amqp.h]: Hey @alanxz, let me know if there's anything more you'd like me to change here.

Comment (manchicken) [reply include/rabbitmq-c/amqp.h]: Hey @alanxz, let me know if there's anything more you'd like me to change here.

**Comment by manchicken** [comment]
@alanxz any updates?

File: include/rabbitmq-c/amqp.h (hunk 1 above)
Comment (alanxz) [reply include/rabbitmq-c/amqp.h]: If all you need is the channel id, then use `amqp_channel_t*` as an output parameter.  

Alternatively consider using an `amqp_channel_t` as an input parameter, then only wait for the confirm on this channel.

File: examples/amqp_ssl_connect.c (hunk 3 above)
Comment (alanxz) [reply examples/amqp_ssl_connect.c:98]: Please send me a separate PR for this change.

File: include/rabbitmq-c/amqp.h [hunk 4]
```diff
@@ -2433,6 +2433,26 @@ AMQP_EXPORT
...
+ *
+ * \param [in] state connection state
+ * \param [in] timeout when waiting for the frame. Passing NULL will result in
+ * blocking behavior
+ * \param [out] pointer to where the envelope details should go
+ * \param [out] pointer to where the ACK details should go
+ * \returns amqp_rpc_reply_t *
+ */
+AMQP_EXPORT
+amqp_rpc_reply_t AMQP_CALL amqp_publisher_confirm_wait(
```
Comment (alanxz) [inline include/rabbitmq-c/amqp.h:2452]: Thinking through this API a bit more we need to consider a couple different scenarios:

1. What does this API do if the connection isn't in publisher-confirm mode? From an API-use perspective, it'd be helpful to return some kind of error if the connection isn't correctly configured.
2. In publisher confirm mode, when a message is published the broker will return one of basic.ack, basic.nack, basic.reject (with channel.close and connection.close as options, but already handled). (see: https://github.com/alanxz/SimpleAmqpClient/blob/500b070465f7759f0bf984ff56b724c06df930ef/src/Channel.cpp#L837-L877). From API-use perspective, it'd be helpful if this function handled all 3 cases, likely this means: this function should have as an out parameter an `amqp_method_number_t`, and a `void*`, which is a library allocated version of one of the previously referenced methods.

**Review by alanxz — CHANGES_REQUESTED** [review]

**Comment by alanxz** [comment]
> @alanxz any updates?

Apologies for the delay, life has been a bit busy lately.

**Comment by manchicken** [comment]
> Apologies for the delay, life has been a bit busy lately.

Don't I know it, friend. Good luck and solidarity.

File: include/rabbitmq-c/amqp.h (hunk 1 above)
Comment (manchicken) [reply include/rabbitmq-c/amqp.h]: I like the pulling the channel as an output parameter, and I'll combine that change with the method out param as well. I think making it filter by channel ID may be outside the scope of what I'm doing here, and the channel out param will allow the caller to easily implement that if they want to.

File: include/rabbitmq-c/amqp.h (hunk 4 above)
Comment (manchicken) [reply include/rabbitmq-c/amqp.h:2452]: Since this new function doesn't own the memory of the decoded payload, it's creating non-fun memory safety issues. I'm thinking of just making a simple union of the different method response types, and having that be an out param instead of the `void*`, and then consolidating all of the out params into one out param.

Something like:

```c
typedef union amqp_publisher_confirm_payload_t_ {
  amqp_basic_ack_t ack;
  amqp_basic_nack_t nack;
  amqp_basic_reject_t reject;
} amqp_publisher_confirm_payload_t;

typedef struct amqp_publisher_confirm_t_ {
  amqp_publisher_confirm_payload_t payload;
  amqp_channel_t channel;
  amqp_method_t method;
} amqp_publisher_confirm_t;
```

Then I can change the function signature to:

```c
amqp_rpc_reply_t AMQP_CALL amqp_publisher_confirm_wait(
    amqp_connection_state_t state, const struct timeval *timeout,
    amqp_publisher_confirm_t *response);
```

That kinda seems better to me, what are your thoughts? This can allow the caller to own all of the memory, and only marginally increases the complexity of this function.

**Comment by manchicken** [comment]
@alanxz I looked at this a bit more and left some notes. Please let me know what you think.

Also, is there any way we could label this PR `hacktoberfest`? 'Tis the season.

File: librabbitmq/amqp_api.c [hunk 5]
```diff
@@ -364,3 +364,61 @@ int amqp_set_rpc_timeout(amqp_connection_state_t state,
   }
   return AMQP_STATUS_OK;
 }
+
+amqp_rpc_reply_t amqp_publisher_confirm_wait(amqp_connection_state_t state,
```
Comment (manchicken) [inline librabbitmq/amqp_api.c:368]: @alanxz: I think I like this a lot better. Do you have any thoughts?

File: librabbitmq/amqp_api.c [hunk 6]
```diff
@@ -364,3 +364,61 @@ int amqp_set_rpc_timeout(amqp_connection_state_t state,
...
+             sizeof(amqp_basic_nack_t));
+      break;
+
+    case AMQP_BASIC_REJECT_METHOD:
+      memcpy(&(result->payload.reject), frame.payload.method.decoded,
+             sizeof(amqp_basic_reject_t));
+      break;
+
+    default:
+      fprintf(stderr, "Got 0x%X for the method, which is «%s».\n",
```
[the remark this answers is not in the document]
Comment (manchicken) [reply librabbitmq/amqp_api.c:411]: Ack! Debug code. Sorry, that's what I get for pushing changes after bedtime. I'll get that.

**Comment by manchicken** [comment]
> #843 should fix the CIFuzz check issue

Did you want me to merge that into my branch?

File: include/rabbitmq-c/amqp.h [hunk 7]
```diff
@@ -2433,6 +2433,42 @@ AMQP_EXPORT
...
+  amqp_basic_reject_t reject; /* basic.reject */
+} amqp_publisher_confirm_payload_t;
+
+/**
+ * Return information from publisher confirm wait
+ **/
+typedef struct amqp_publisher_confirm_t_ {
+  amqp_publisher_confirm_payload_t payload; /* The response payload */
+  amqp_channel_t channel; /* The channel where the confirmation was received */
+  amqp_method_number_t method; /* The method which was received */
```
[the remark this answers is not in the document]
Comment (manchicken) [reply include/rabbitmq-c/amqp.h:2451]: I added this to the function's docs.

File: examples/amqp_confirm_select.c [hunk 8]
```diff
@@ -0,0 +1,192 @@
+// Copyright 2007 - 2021, Alan Antonuk and the rabbitmq-c contributors.
+// SPDX-License-Identifier: mit
+
+#include <inttypes.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
```
[the remark this answers is not in the document]
Comment (manchicken) [reply examples/amqp_confirm_select.c]: Good catch; I just moved it.

[the remark this answers is not in the document]
Comment (manchicken) [reply examples/amqp_confirm_select.c]: Does this need `windows.h` as well? I don't have a Windows box to try compiling on.

**Comment by manchicken** [comment]
@alanxz Any updates? I'd like to see if I can get things moving along soon so I can start my work on the Perl side of what I'm doing with this.

File: examples/amqp_confirm_select.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (alanxz) [reply examples/amqp_confirm_select.c]: I think you'll need to #include <windows.h> in order to get the `sleep` function.

[the remark this answers is not in the document]
Comment (manchicken) [reply examples/amqp_confirm_select.c]: I'm gonna pull the sleep function. No need for the complexity.

File: include/rabbitmq-c/amqp.h [hunk 9]
```diff
@@ -2433,6 +2433,54 @@ AMQP_EXPORT
...
+typedef struct amqp_publisher_confirm_t_ {
+  amqp_publisher_confirm_payload_t payload; /* The response payload; check the `method` value to see which value you should use in the union */
+  amqp_channel_t channel; /* The channel where the confirmation was received */
+  amqp_method_number_t method; /* The method which was received */
+} amqp_publisher_confirm_t;
+
+/**
+ * amqp_publisher_confirm_wait
+ *
+ * Wait for a publisher confirm when the connection is in select mode.
```
Comment (alanxz) [inline include/rabbitmq-c/amqp.h:2457]: nit: select mode is enabled per channel (please update line 2457)

**Review by alanxz — APPROVED** [review]
Two minor things, otherwise this is good to go, thanks for your patience.

**Comment by manchicken** [comment]
🥳

**Comment by manchicken** [comment]
@alanxz Do you know when this will be released? Not trying to rush you, just trying to time things better.

**Comment by alanxz** [comment]
@manchicken I'll try and roll a release in the next few weeks.

**Comment by manchicken** [comment]
Thanks!~ Mike StemleOn Oct 25, 2024, at 08:46, Alan Antonuk ***@***.***> wrote:﻿
@manchicken I'll try and roll a release in the next few weeks.

**Comment by alanxz** [comment]
I've released v0.15.0
